File: src/net/opgenorth/yeg/model/SortByDistanceFromLocation.java
package net.opgenorth.yeg.model;

import android.location.Location;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class SortByDistanceFromLocation implements IHistoricalBuildingSorter {
	private Location _location;

	public SortByDistanceFromLocation(Location location) {
		_location = location;
	}

	@Override
	public List<HistoricalBuilding> sortList(List<HistoricalBuilding> buildings) {
		ArrayList<BuildingAndLocationWrapper> list = new ArrayList<BuildingAndLocationWrapper>(buildings.size());
		for (HistoricalBuilding building : buildings) {
			list.add(new BuildingAndLocationWrapper(building, _location));
		}

		Collections.sort(list);

		ArrayList<HistoricalBuilding> sortedList = new ArrayList<HistoricalBuilding>(buildings.size());
		for (BuildingAndLocationWrapper building : list) {
			sortedList.add(building.getHistoricalBuilding());
		}
		return sortedList;
	}
}

File: dotnet/src/yegbuildings/Model/SortByDistanceFromLocation.cs
using System.Collections.Generic;
using System.Linq;
using Android.Locations;
using Net.Opgenorth.Yeg.Buildings.Util;

namespace Net.Opgenorth.Yeg.Buildings.model
{
    public class SortByDistanceFromLocation : IBuildingSorter
    {
        private readonly Location _location;

        public SortByDistanceFromLocation(Location location)
        {
            _location = location;
        }

        public Location Location
        {
            get { return _location; }
        }

        public IList<Building> Sort(IEnumerable<Building> buildings)
        {
            if (_location == null)
            {
                return buildings.ToList();
            }

            var locationToBuildings = from building in buildings
                                      select new
                                                 {
                                                     Building = building,
                                                     Distance = _location.DistanceTo(building)
                                                 };
            var sortedList = from l2b in locationToBuildings
                             orderby l2b.Distance
                             select l2b.Building;

            return sortedList.ToList();
        }
    }
}